src/transports/ipc.rs

```diff
@@ -0,0 +1,185 @@
+use crate::{helpers, Error, RequestId, Result, Transport};
+use jsonrpc_core as rpc;
+use std::{
+    collections::BTreeMap,
+    path::Path,
+    pin::Pin,
+    sync::{atomic::AtomicUsize, Arc},
+    task::{Context, Poll},
+};
+use tokio::{
+    io::{reader_stream, AsyncWriteExt},
+    net::{unix, UnixStream},
+    stream::StreamExt,
+    sync::{mpsc, oneshot},
+};
+
+#[derive(Debug, Clone)]
+pub struct Ipc {
+    id: Arc<AtomicUsize>,
+    messages_tx: mpsc::UnboundedSender<TransportMessage>,
+}
+
+impl Ipc {
+    pub async fn new<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let id = Arc::new(AtomicUsize::new(1));
+        let (messages_tx, messages_rx) = mpsc::unbounded_channel();
+
+        let socket = UnixStream::connect(path).await?;
+        tokio::spawn(run_server(socket, messages_rx));
+
+        Ok(Ipc { id, messages_tx })
+    }
+}
+
+impl Transport for Ipc {
+    type Out = Response;
+
+    fn prepare(&self, method: &str, params: Vec<rpc::Value>) -> (crate::RequestId, rpc::Call) {
+        let id = self.id.fetch_add(1, std::sync::atomic::Ordering::AcqRel);
+        let request = helpers::build_request(id, method, params);
+        (id, request)
+    }
+
+    fn send(&self, id: RequestId, request: rpc::Call) -> Self::Out {
+        let (response_tx, response_rx) = oneshot::channel();
+
+        let message = TransportMessage {
+            request: rpc::Request::Single(request),
+            response_tx,
+            id,
+        };
+
+        Response(self.messages_tx.send(message).map(|()| response_rx).map_err(Into::into))
+    }
+}
+
+pub struct Response(Result<oneshot::Receiver<rpc::Value>>);
+
+impl futures::Future for Response {
+    type Output = Result<rpc::Value>;
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match &mut self.0 {
+            Err(err) => Poll::Ready(Err(err.clone())),
+            Ok(ref mut rx) => {
+                let value = ready!(futures::Future::poll(Pin::new(rx), cx))?;
+                Poll::Ready(Ok(value))
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+struct TransportMessage {
+    id: RequestId,
+    request: rpc::Request,
+    response_tx: oneshot::Sender<rpc::Value>,
+}
+
+async fn run_server(unix_stream: UnixStream, messages_rx: mpsc::UnboundedReceiver<TransportMessage>) -> Result<()> {
+    let (socket_reader, mut socket_writer) = unix_stream.into_split();
+    let mut pending_response_txs = BTreeMap::default();
+
+    let mut socket_reader = reader_stream(socket_reader);
+    let mut messages_rx = messages_rx.fuse();
+    let mut read_buffer = vec![];
+
+    loop {
+        tokio::select! {
+            message = messages_rx.next() => match message {
```

Comment:
The `match` here could be `if let Some(message) = message {...}`

Reply:
Done in 1788be6f69e1a7864572b35f1c396091cfdfd014